lib/TH/generic/THTensorMath.c

```diff
@@ -1754,22 +1754,61 @@ void THTensor_(topk)(THTensor *rt_, THLongTensor *ri_, THTensor *t, long k, int
   long sliceSize = THTensor_(size)(t, dim);
   THArgCheck(k > 0 && k <= sliceSize, 2, "k not in range for dimension");
 
-  /* Just implement in terms of sort and narrow for now */
-  THTensor* tmpResults = THTensor_(new)();
-  THLongTensor* tmpIndices = THLongTensor_new();
+  THTensor *tmpResults = THTensor_(new)();
+  THTensor_(resize1d)(tmpResults, sliceSize);
+  real *tmp__data = THTensor_(data)(tmpResults);
 
-  THLongStorage* topKSize = THTensor_(newSizeOf)(t);
+  THLongTensor *tmpIndices = THLongTensor_new();
+  THLongTensor_resize1d(tmpIndices, sliceSize);
+  long *tmpi__data = THLongTensor_data(tmpIndices);
+
+  THLongStorage *topKSize = THTensor_(newSizeOf)(t);
   THLongStorage_set(topKSize, dim, k);
   THTensor_(resize)(rt_, topKSize, NULL);
   THLongTensor_resize(ri_, topKSize, NULL);
   THLongStorage_free(topKSize);
 
-  THTensor_(sort)(tmpResults, tmpIndices, t, dim, dir);
-  THTensor_(narrow)(tmpResults, NULL, dim, 0, k);
-  THLongTensor_narrow(tmpIndices, NULL, dim, 0, k);
+  if (dir) {
```

Comment:
Well it's actually more or less the same in terms of the algorithm. The only other idea that I can think of is to construct a heap for each slice and then extract k elements from it. The upside is that you get them sorted for free. I did some search and apparently (as [in here](http://blog.mischel.com/2011/10/25/when-theory-meets-practice/)) it might be faster than `quickselect` for small k (also, this post didn't include the sorting - it was only a one element selection). I think that the only way to check what's better is to benchmark :smiley: